qualang_tools/bakery/readme.md

```diff
@@ -0,0 +1,273 @@
+# Waveform baking
+This library introduces a new framework for creating arbitrary waveforms and storing them in a usual configuration file. 
+The idea is to simplify the writing of pulse samples that would usually be declared as incompatible with respect to QUA restrictions for uploading the samples to the OPX. A nice advantage of using this tool is to embed into one single waveform a series of instructions that allows program memory preservation.
+
+Waveform baking is embedded into a new context manager, declared prior to the QUA program, that takes two inputs: 
+
+- the configuration dictionary (the same used to initialize a Quantum Machine instance),
+
+- a padding method: to be chosen between: “right”, “left”, “symmetric_l”, “symmetric_r”.  This string indicates how the should samples be filled up with 0s when they do not correspond to a usual sample (that is if sample length is not a multiple of 4 or if it is shorter than 16 ns). 
+
+    -  “right” setting is the default setting and pads 0s at the end of the baked sample to insert a QUA compatible version in the original configuration file
+
+    - “left” pads 0s before the baked sample
+
+    - “symmetric_l” pads 0s symmetrically before and after the baked sample, putting one more 0 before it in case the baked sample length is odd
+
+    - “symmetric_r' pads 0s symmetrically before and after the baked sample , putting one more 0 after it in case the baked sample length is odd
+
+Declaration is done before the QUA program as follows: 
+
+```
+with baking(config, padding_method = "symmetric_r") as b:
+  b.align("qe1", "qe2", "qe3")
+  b.frame_rotation(0.78, "qe2")
+  b.ramp(amp=0.3, duration=9, qe="qe1")
+```
+
+This context manager does not return any output, its execution results in an edition of the configuration file provided as input to add:
+- an operation for each quantum element involved within the baking context manager
+- an associated pulse
+- an associated waveform (set of 2 waveforms for a mixedInputs quantum element) containing  sample(s) issued from concatenation of operations indicated in the context manager.
+  
+The baking also carries advantages for an efficient program memory management,
+in the sense that there are less commands and different samples to be stored in memory to execute the program.
+
+# **How can I add operations inside the baking context manager?**
+
+The logic behind the baking context manager is to stay as close as possible to the way we would write play statements within a QUA program. For instance, commands like frame_rotation, reset_phase, ramp, wait and align are all replicated  within the context manager. 
+
+For playing custom operations, from an arbitrary shaped waveform the procedure goes as follows:

```

Comment:
For playing custom operations, from an arbitrary shaped waveform the procedure goes as follows:

------

The procedure for playing customer operations is as follows: